File: 백준/Gold/5014. 스타트링크/스타트링크.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Queue struct {
	Node []int
}

func (q *Queue) Enqueue(n int) {
	q.Node = append(q.Node, n)
}
func (q *Queue) Dequeue() (n int) {
	if len(q.Node) <= 0 {
		return -1
	} else {
		n = q.Node[0]
		if len(q.Node) == 1 {
			q.Node = []int{}
		} else {
			q.Node = q.Node[1:]
		}
	}
	return n
}

func (q *Queue) Size() int {
	return len(q.Node)
}

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter(os.Stdout)
	defer writer.Flush()
	var f, s, g, u, d int
	fmt.Fscanf(reader, "%d %d %d %d %d\n", &f, &s, &g, &u, &d)
	var visited = make([]bool, f+1)
	var count = make([]int, f+1)

	var q = new(Queue)
	q.Enqueue(s)
	visited[s] = true
	for q.Size() > 0 {
		now := q.Dequeue()
		if now == g {
			break
		}
		if 0 < now-d && now-d <= f && visited[now-d] == false {
			visited[now-d] = true
			q.Enqueue(now - d)
			count[now-d] = count[now] + 1
		}
		if 0 < now+u && now+u <= f && visited[now+u] == false {
			visited[now+u] = true
			q.Enqueue(now + u)
			count[now+u] = count[now] + 1
		}
	}
	if count[g] > 0 || s == g {
		fmt.Fprintf(writer, "%d\n", count[g])
	} else {

		fmt.Fprintf(writer, "%s\n", "use the stairs")
	}
}

File: 백준/Gold/5014. 스타트링크/스타트링크.py
import sys 
from collections import deque
input = sys.stdin.readline
sys.setrecursionlimit(10**6)
def bfs():
    q = deque()
    q.append(S)
    visited[S]=True
    while q:
        now = q.popleft()
        if now ==G:
            return count[G]
        for v in (now-D, now+U):
            if 0<v<=F and visited[v]==False:
                visited[v]=True
                q.append(v)
                count[v]=count[now]+1
        
    if count[G]<=0:
        return "use the stairs"
F,S,G,U,D = map(int, input().split())
#총 F 목적 G 현재 S 
visited = [False for _ in range(F+1)]
count = [0 for _ in range(F+1)]
print(bfs())